pandas/_libs/src/ujson/python/objToJSON.c

```diff
@@ -238,8 +238,10 @@ static PyObject *get_values(PyObject *obj) {
             PyErr_Clear();
         } else if (PyObject_HasAttrString(values, "__array__")) {
             // We may have gotten a Categorical or Sparse array so call np.array
+            PyObject *array_values = PyObject_CallMethod(values, "__array__",
```

Comment:
Do you know how the memory management work here? I'm not entirely sure that is safe to reassign to `values` after decrementing. It may be happen-stance that this improves the odds of delaying garbage collection. Maybe we can just return array_values here directly?

Reply:
In C there is no Python style reference counting. I must say that I assume that the Python object returned by `PyObject_CallMethod` has at least a reference count of 1, so that it does not get deallocated.
The next line decrements the refcount of the original `values` object, which in this case will reduce it to 0, causing it to be freed immediately. This was the problem causing the segfault.
This fix delays the destruction of the original `values` object and as the `array_values` object should have a refcount high enough not to be destroyed, this works.

Reply:
Also in case `array_values` would be equal to NULL, this if statement on line 252 (new situation) would not be executed if it would it would be returned directly.

Reply:
> I must say that I assume that the Python object returned by `PyObject_CallMethod` has at least a reference count of 1, so that it does not get deallocated.

Yep typicaly calls`PyObject_*` functions will give you ownership of the reference to an object, which is pretty similar to +1 on a refcount. 

> The next line decrements the refcount of the original `values` object, which in this case will reduce it to 0, causing it to be freed immediately.

Could be wrong but I think it gets freed on the next garbage collector run, not necessarily immediately when the count reaches zero. Would definitely be safer here to return rather than re-using the variable if we can

Reply:
not for this PR, but this part of the thread re-ups my desire to move this logic out of C

Reply:
The Python documentation on reference counting is definitely a good resource. Worth a read:

https://docs.python.org/3/c-api/intro.html#reference-count-details

I think this code would be very difficult to port to Cython maybe not even worth it, but of course anything possible with time and effort

Reply:
When the reference count reaches zero, a python object gets freed (see [Py_DECREF](https://docs.python.org/3/c-api/refcounting.html#c.Py_DECREF)). Garbage collection is used to free up objects which are held in circular references that are no longer referenced from any other objects.
So there is nothing to be gained by returning immediately with respect to influencing the reference count.

Reply:
Makes sense. Thanks for clarifying